Test/ServerTelemetryChannel.Test/Shared.Tests/Implementation/BackoffLogicManagerTest.cs

```diff
@@ -0,0 +1,248 @@
+
+    public class BackoffLogicManagerTest
+    {
+        [TestClass]
+        public class GetBackendResponse
+        {
+            public void ReturnNullIfArgumentIsNull()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                Assert.Null(manager.GetBackendResponse(null));
+            }
+
+            public void ReturnNullIfArgumentEmpty()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                Assert.Null(manager.GetBackendResponse(string.Empty));
+            }
+
+            public void IfContentCannotBeParsedNullIsReturned()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                Assert.Null(manager.GetBackendResponse("ab}{"));
+            }
+
+            public void IfContentIsUnexpectedJsonNullIsReturned()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                Assert.Null(manager.GetBackendResponse("[1,2]"));
+            }
+
+            public void BackendResponseIsReturnedForCorrectContent()
+            {
+                string content = BackendResponseHelper.CreateBackendResponse(100, 1, new[] {"206"}, 84);
+
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                var backendResponse = manager.GetBackendResponse(content);
+
+                Assert.Equal(1, backendResponse.ItemsAccepted);
+                Assert.Equal(100, backendResponse.ItemsReceived);
+                Assert.Equal(1, backendResponse.Errors.Length);
+                Assert.Equal(84, backendResponse.Errors[0].Index);
+                Assert.Equal(206, backendResponse.Errors[0].StatusCode);
+                Assert.Equal("Explanation", backendResponse.Errors[0].Message);
+            }
+        }
+
+        [TestClass]
+        public class ScheduleRestore
+        {
+            [TestMethod]
+            public void NoErrorDelayIsSameAsSlotDelay()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                manager.ScheduleRestore(string.Empty, () => null);
+                Assert.Equal(TimeSpan.FromSeconds(10), manager.CurrentDelay);
+            }
+
+            [TestMethod]
+            public void FirstErrorDelayIsSameAsSlotDelay()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero);
+                manager.ConsecutiveErrors++;
+                manager.ScheduleRestore(string.Empty, () => null);
+                Assert.Equal(TimeSpan.FromSeconds(10), manager.CurrentDelay);
+            }
+
+            [TestMethod]
+            public void UpperBoundOfDelayIsMaxDelay()
+            {
+                var manager = new BackoffLogicManager(TimeSpan.Zero) { ConsecutiveErrors = int.MaxValue };
+                manager.ScheduleRestore(string.Empty, () => null);
+                Assert.InRange(manager.CurrentDelay, TimeSpan.FromSeconds(10), TimeSpan.FromSeconds(3600));
```

Comment:
I would not want to test if product and test code uses same constant. If we change 3600 in product code I want someone to change this test so that he is doing this consciously. 

---
In reply to: [69628722](https://github.com/Microsoft/ApplicationInsights-dotnet/pull/248#discussion_r69628722) [](ancestors = 69628722)